hail/src/main/scala/is/hail/expr/types/physical/PNDArray.scala

```diff
@@ -33,74 +54,66 @@ final case class PNDArray(elementType: PType, nDims: Int, override val required:
 
   override def fundamentalType: PType = representation.fundamentalType
 
-  def numElements(shape: Code[Long], mb: MethodBuilder): Code[Long] = {
-    def getShapeAtIdx(idx: Int): Code[Long] = Region.loadLong(this.representation.fieldType("shape").asInstanceOf[PTuple]
-      .loadField(shape, idx))
-
-    if (nDims == 0) {
-      0L
-    }
-    else {
-      Array.range(0, nDims).foldLeft(const(1L)) { (prod, idx) => prod * getShapeAtIdx(idx) }
-    }
+  def numElements(shape: Array[Code[Long]], mb: MethodBuilder): Code[Long] = {
+      Array.range(0, nDims).foldLeft(const(1L)) { (prod, idx) => prod * shape(idx) }
```

Comment:
now that we're using shape arrays this is just `shape.foldLeft(const(1L), _ * _)`